File: src/rb/PhantasyIslandRubyRemoteControl/image_receiver_mook.rb
require 'thread'

class ImageInfo
  attr_accessor :img, :id, :total_count, :progress_count, :ok

  def initialize(img:, id:, total_count:, progress_count: 0, ok: false)
    @img = img
    @id = id
    @total_count = total_count
    @progress_count = progress_count
    @ok = ok
  end
end

class ImageReceiver
  attr_accessor :airplane, :image_instance, :now_loading_id, :user_receive_callback, :user_progress_callback

  def initialize(airplane)
    @airplane = airplane
    @image_instance = nil
    @cmd_id_counter = 1
    @now_loading_id = 0
    @lock = Mutex.new
    @mook_time = 3
  end

  def send_cap_image(user_receive_callback = nil, user_progress_callback = nil)
    @user_receive_callback = user_receive_callback
    @user_progress_callback = user_progress_callback

    Thread.new do
      mook_receive_thread
    end
  end

  def mook_receive_thread
    @lock.synchronize do
      @cmd_id_counter += 1
      @now_loading_id = @cmd_id_counter
      @image_instance = ImageInfo.new(
        img: @airplane.get_camera_down_img,
        id: @cmd_id_counter,
        total_count: @mook_time * 100
      )
    end

    while @image_instance.progress_count < @image_instance.total_count
      sleep(0.01)
      @lock.synchronize do
        break if @image_instance.id != @now_loading_id
        @image_instance.progress_count += 1
      end

      if @user_progress_callback
        @user_progress_callback.call(@image_instance.progress_count, @image_instance.total_count)
      end
    end

    @lock.synchronize do
      if @image_instance.id == @now_loading_id
        @image_instance.ok = true
      end
    end

    if @image_instance.id == @now_loading_id && @image_instance.progress_count >= @image_instance.total_count && @user_receive_callback
      @user_receive_callback.call(@image_instance.img)
    end
  end

  def get_latest_image
    @lock.synchronize do
      return nil if @image_instance.nil?
      @image_instance.ok ? @image_instance.img : nil
    end
  end

  def get_transfer_progress
    @lock.synchronize do
      return nil if @image_instance.nil?
      @image_instance.progress_count
    end
  end

  def is_transfer_in_progress
    @lock.synchronize do
      return false if @image_instance.nil?
      !@image_instance.ok
    end
  end
end

File: src/kt/PhantasyIslandKotlinRemoteControl/ImageReceiverMook.kt
package PhantasyIslandKotlinRemoteControl

import java.awt.image.BufferedImage
import java.util.concurrent.locks.ReentrantReadWriteLock
import kotlin.concurrent.withLock
import kotlin.concurrent.thread

data class ImageInfo(
    val img: BufferedImage?,
    val id: Int,
    val totalCount: Int,
    var progressCount: Int = 0,
    var ok: Boolean = false
)

class ImageReceiver(private val airplane: AirplaneCore) {

    // 任何时候只存在一张图片
    private var imageInstance: ImageInfo? = null
    private var cmdIdCounter: Int = 1
    private var nowLoadingId: Int = 0
    private val lock = ReentrantReadWriteLock()

    var userReceiveCallback: ((BufferedImage?) -> Unit)? = null
    var userProgressCallback: ((Int, Int) -> Unit)? = null

    val mookTime = 3

    fun sendCapImage(
        userReceiveCallback: ((BufferedImage?) -> Unit)? = null,
        userProgressCallback: ((Int, Int) -> Unit)? = null
    ) {
        this.userReceiveCallback = userReceiveCallback
        this.userProgressCallback = userProgressCallback

        thread {
            mookReceiveThread()
        }
    }

    private fun mookReceiveThread() {
        lock.writeLock().withLock {
            cmdIdCounter += 1
            nowLoadingId = cmdIdCounter
            imageInstance = ImageInfo(
                img = airplane.getCameraDownImg(),
                id = cmdIdCounter,
                totalCount = mookTime * 100 // 模拟 3 秒，每 10ms 更新一次
            )
        }

        while (true) {
            val (currentProgress, currentTotal, currentId) = lock.readLock().withLock {
                val inst = imageInstance ?: return@thread
                Triple(inst.progressCount, inst.totalCount, inst.id)
            }

            if (currentProgress >= currentTotal || currentId != nowLoadingId) {
                break
            }

            Thread.sleep(10)

            lock.writeLock().withLock {
                if (imageInstance?.id == nowLoadingId) {
                    imageInstance?.progressCount = (imageInstance?.progressCount ?: 0) + 1
                }
            }

            userProgressCallback?.invoke(currentProgress + 1, currentTotal)
        }

        var finalImg: BufferedImage? = null
        lock.writeLock().withLock {
            if (imageInstance?.id == nowLoadingId) {
                imageInstance?.ok = true
                finalImg = imageInstance?.img
            }
        }

        if (finalImg != null && nowLoadingId != 0) {
            userReceiveCallback?.invoke(finalImg)
        }
    }

    fun getLatestImage(): BufferedImage? {
        return lock.readLock().withLock {
            if (imageInstance?.ok == true) imageInstance?.img else null
        }
    }

    fun getTransferProgress(): Int? {
        return lock.readLock().withLock {
            imageInstance?.progressCount
        }
    }

    fun isTransferInProgress(): Boolean {
        return lock.readLock().withLock {
            imageInstance != null && imageInstance?.ok == false
        }
    }
}
